### PR TITLE
refactor(core): Modernize logger service

### PR DESCRIPTION
**Pull request by ivov**
Clean up logger service, port logging config, upgrade logging lib, write tests.

This is setup for introducing [scoped debug logging](https://linear.app/n8n/issue/PAY-2016).

This PR looks big but the bulk of it is import changes. Please focus on:

- `packages/cli/logging/`
- `packages/@n8n/config/src/configs/logging.config.ts`
- `winston@3.8.2` → `winston@3.14.2`
- Three spots where ported config is being consumed
	- `packages/cli/src/posthog/index.ts`
	- `packages/cli/src/services/ai-assistant.service.ts`
	- `packages/cli/src/services/frontend.service.ts`

### CODE REVIEW

File: packages/@n8n/config/src/configs/logging.config.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	@Env('N8N_LOG_LEVEL')
+	level: 'error' | 'warn' | 'info' | 'debug' | 'silent' = 'info';
+
+	/**
+	 * Where to output logs to. Options are: `console` or `file` or both in a comma separated list.
+	 *
+	 * @example `N8N_LOG_OUTPUT=console,file` will output to both console and file.
+	 */
+	@Env('N8N_LOG_OUTPUT')
+	output: string = 'console';
```
Comment (netroy) [inline packages/@n8n/config/src/configs/logging.config.ts:42]: Just how we created `JsonStringArray`, we could create a `StringArray` class
```typescript
class StringArray<T extends string> extends Array<T> {
	constructor(str: string) {
		super();
		const parsed = str.split(',') as StringArray<T>;
		return isStringArray(parsed) ? parsed : [];
	}
}
```

and use that here like

```suggestion
	output: StringArray<'console' | 'file'>  = ['console'];
```

File: packages/cli/src/logging/logger.service.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+			if (this.outputs.includes('console')) this.setConsoleTransport();
+			if (this.outputs.includes('file')) this.setFileTransport();
+		}
+
+		LoggerProxy.init(this);
+	}
+
+	get outputs() {
+		return this.globalConfig.logging.output.split(',').map((o) => o.trim());
+	}
```
Comment (netroy) [inline packages/cli/src/logging/logger.service.ts:46]: If we move the string parsing in the config object, we don't need to do this here.

**Review by netroy — CHANGES_REQUESTED** [review]
besides the string array thing, lgtm

**Review by despairblue — COMMENTED** [review]
Looks good. I left some questions for clarification.

File: packages/cli/src/logging/logger.service.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	private setLevel() {
+		const { levels } = this.internalLogger;
+
+		for (const logLevel of LOG_LEVELS) {
+			if (levels[logLevel] > levels[this.level]) {
+				// winston defines `{ error: 0, warn: 1, info: 2, debug: 5 }`
+				// so numerically higher (less severe) log levels become no-op
+				Object.defineProperty(this, logLevel, { value: noOp });
+			}
+		}
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/cli/src/logging/logger.service.ts:71]: more context [here](https://github.com/n8n-io/n8n/pull/7479)

File: packages/@n8n/config/src/configs/logging.config.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	fileCountMax: number = 100;
+
+	/** Max size (in MiB) for each log file. Only for `file` log output. */
+	@Env('N8N_LOG_FILE_SIZE_MAX')
+	fileSizeMax: number = 16;
+
+	/** Location of the log files inside `~/.n8n`. Only for `file` log output. */
+	@Env('N8N_LOG_FILE_LOCATION')
+	location: string = 'logs/n8n.log';
+}
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/@n8n/config/src/configs/logging.config.ts:24]: Agreed, but that would unfortunately be a breaking change. Maybe for v2?

File: packages/cli/src/logging/logger.service.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+						}),
+					)
+				: winston.format.printf(({ message }: { message: string }) => message);
+
+		this.internalLogger.add(new winston.transports.Console({ format }));
+	}
+
+	private toPrintable(metadata: unknown) {
+		if (isObjectLiteral(metadata) && Object.keys(metadata).length > 0) {
+			return ' ' + JSON.stringify(inspect(metadata));
```
[the remark this answers is not in the document]
Comment (ivov) [reply packages/cli/src/logging/logger.service.ts]: Good find!

[the remark this answers is not in the document]
Comment (ivov) [reply packages/cli/src/logging/logger.service.ts]: [2e45ab4](https://github.com/n8n-io/n8n/pull/11031/commits/2e45ab4b6f307cd97eba0c75c823d5c278f32449)

File: packages/@n8n/config/src/configs/logging.config.ts (hunk 1 above)
Comment (ivov) [reply packages/@n8n/config/src/configs/logging.config.ts]: [d3c794a](https://github.com/n8n-io/n8n/pull/11031/commits/d3c794a6192ea03937b61a281d5460dc3f6aee8d)

File: packages/cli/src/logging/logger.service.ts (hunk 2 above)
Comment (ivov) [reply packages/cli/src/logging/logger.service.ts]: [d3c794a](https://github.com/n8n-io/n8n/pull/11031/commits/d3c794a6192ea03937b61a281d5460dc3f6aee8d)

File: packages/cli/src/logging/logger.service.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ivov) [reply packages/cli/src/logging/logger.service.ts]: [77a4468](https://github.com/n8n-io/n8n/pull/11031/commits/77a4468978720d8297c1b413665fa8dc1fc784e4)

[the remark this answers is not in the document]
Comment (netroy) [reply packages/cli/src/logging/logger.service.ts]: TBH, if we can skip the `callsites` call in some other way that makes this code less confusing, I'd prefer that as well.

**Review by despairblue — APPROVED** [review]

File: packages/cli/src/logging/logger.service.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (despairblue) [reply packages/cli/src/logging/logger.service.ts]: I was thinking about the log methods short-circuiting. I'd prefer that to monkey patching.

**Review by netroy — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.62.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.62.1)

**Pull request merged**